src/libraries/System.Security.Cryptography.Xml/src/System/Security/Cryptography/Xml/CanonicalXmlDocument.cs

```diff
@@ -61,54 +61,56 @@ public void WriteHash(HashAlgorithm hash, DocPosition docPos, AncestralNamespace
         }
 
-        public override XmlAttribute CreateAttribute(string prefix, string localName, string namespaceURI)
+        public override XmlAttribute CreateAttribute(string? prefix, string localName, string? namespaceURI)
         {
             return new CanonicalXmlAttribute(prefix, localName, namespaceURI, this, _defaultNodeSetInclusionState);
         }
 
-        protected override XmlAttribute CreateDefaultAttribute(string prefix, string localName, string namespaceURI)
+        protected override XmlAttribute CreateDefaultAttribute(string? prefix, string localName, string? namespaceURI)
         {
             return new CanonicalXmlAttribute(prefix, localName, namespaceURI, this, _defaultNodeSetInclusionState);
         }
 
-        public override XmlText CreateTextNode(string text)
+        public override XmlText CreateTextNode(string? text)
         {
             return new CanonicalXmlText(text, this, _defaultNodeSetInclusionState);
         }
 
-        public override XmlWhitespace CreateWhitespace(string prefix)
+        public override XmlWhitespace CreateWhitespace(string? prefix)
         {
-            return new CanonicalXmlWhitespace(prefix, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlWhitespace(prefix!, this, _defaultNodeSetInclusionState);
         }
 
-        public override XmlSignificantWhitespace CreateSignificantWhitespace(string text)
+        public override XmlSignificantWhitespace CreateSignificantWhitespace(string? text)
         {
-            return new CanonicalXmlSignificantWhitespace(text, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlSignificantWhitespace(text!, this, _defaultNodeSetInclusionState);
         }
 
         public override XmlProcessingInstruction CreateProcessingInstruction(string target, string data)
         {
             return new CanonicalXmlProcessingInstruction(target, data, this, _defaultNodeSetInclusionState);
         }
 
-        public override XmlComment CreateComment(string data)
+        public override XmlComment CreateComment(string? data)
         {
-            return new CanonicalXmlComment(data, this, _defaultNodeSetInclusionState, _includeComments);
+            return new CanonicalXmlComment(data!, this, _defaultNodeSetInclusionState, _includeComments);
         }
 
         public override XmlEntityReference CreateEntityReference(string name)
         {
             return new CanonicalXmlEntityReference(name, this, _defaultNodeSetInclusionState);
         }
 
-        public override XmlCDataSection CreateCDataSection(string data)
+        public override XmlCDataSection CreateCDataSection(string? data)
         {
-            return new CanonicalXmlCDataSection(data, this, _defaultNodeSetInclusionState);
+            return new CanonicalXmlCDataSection(data!, this, _defaultNodeSetInclusionState);
```

Comment:
Why are these `!`s necessary?  It looks like that first argument to all these CanonicalXx ctors is just passed down to the base constructor, which is annotated as `string?`... should all of these be `string?` as well?

Reply:
`XmlDocument` declares this as `public virtual XmlCDataSection CreateCDataSection(string? data)`, so, from what I can see, this override must have the same signature.

The constructor for `CanonicalXmlCDataSection` looks like this:
```csharp
public CanonicalXmlCDataSection(string data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
```

So, it looks like the `!`s are necessary (at least to me, I stand to be corrected by you shortly :)  )

Reply:
@stephentoub - just directly pinging you as these comments are not appearing on the 'conversation' tab. I think this is because I had to rebase due to changes on main.

Reply:
I'm not following. XmlCDataSection has this:
```C#
        protected internal XmlCDataSection(string? data, XmlDocument doc) : base(data, doc)
        {
        }
```
CanonicalXmlCDataSection has this:
```
        public CanonicalXmlCDataSection(string data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
        {
            _isInNodeSet = defaultNodeSetInclusionState;
        }
```
and this PR is providing the nullable annotations.  So shouldn't this PR be making it:
```C#
        public CanonicalXmlCDataSection(string? data, XmlDocument doc, bool defaultNodeSetInclusionState) : base(data, doc)
        {
            _isInNodeSet = defaultNodeSetInclusionState;
        }
```
at which point the `!` on the first arugment to that ctor isn't needed?